presto-main/src/main/java/com/facebook/presto/operator/PartitionedConsumption.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.log.Logger;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.IntConsumer;
+import java.util.function.IntFunction;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.Futures.allAsList;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Coordinates consumption of a resource by multiple consumers
+ * in a partition-by-partition manner.
+ * <p>
+ * The partitions to consume can be obtained by consumers using the
+ * {@code Iterable<Partition<T>> getPartitions()} method.
+ * <p>
+ * A {@code Partition} is only loaded after at least one consumer has called
+ * {@code Partition#load()} on it and all the {@code consumersCount} consumers
+ * have {@code release()}-d the previous {@code Partition} (if any).
+ * <p>
+ * The loaded object is {@code close()}-d when all the consumers have released
+ * its {@code Partition} and thus must implement {@code Closable}.
+ * <p>
+ * The partitions contents are loaded using the {@code Function<Integer, CompletableFuture<T>> loader} passed
+ * upon construction. The integer argument in the loader function is the number of the partition to load.
+ * <p>
+ * The partition number can be accessed using {@code Partition#number()} and - for a given partition -
+ * it takes the value of the respective element of {@code partitionNumbers} passed upon construction.
+ *
+ * @param <T> type of the object loaded for each {@code Partition}. Must be {@code Closable}.
```

Comment:
there are no longer bounds on T type. Also, maybe add a note on the ` IntConsumer disposer`